src/tunnel.cc

```diff
@@ -625,6 +661,20 @@ TunnelStateData::Connection::write(const char *b, int size, AsyncCall::Pointer &
     Comm::Write(conn, b, size, callback, free_func);
 }
 
+void
+TunnelStateData::Connection::setCloseHandler(CLCB *handler, TunnelStateData *tunnelState)
+{
+    assert(handler);
+    assert(tunnelState);
+    if (!Comm::IsConnOpen(conn)) {
```

Comment:
Why is this check needed? It seems to contradict what the method is promising to do. Sooner or later the caller may incorrectly rely on the closer being set. If the check is necessary, perhaps the method should be renamed to setCloseHandlerIfOpen()? Otherwise, this condition may become an assertion (or we could rely on a similar assertion in comm_remove/add_close_handler() if there is one in those functions).